src/group_theory/free_product.lean

```diff
@@ -327,4 +328,19 @@ instance : decidable_eq (free_product M) := word.equiv.decidable_eq
 
 end word
 
+/-- The free product of free groups is itself a free group -/
+@[simps]
+instance {ι : Type*} (G : ι → Type*) [∀ i, group (G i)] [hG : ∀ i, is_free_group (G i)] :
+  is_free_group (free_product G) :=
+{ generators := Σ i, is_free_group.generators (G i),
+  of := λ x, free_product.of (is_free_group.of x.2),
+  unique_lift' :=
+  begin
+    introsI X _ f,
+    refine ⟨free_product.lift (λ i, is_free_group.lift (λ x, f ⟨i, x⟩)), _ ⟩,
+    split,
+    { simp, },
+    { intros g hfg, ext i x, simpa using hfg ⟨i, x⟩, }
```

Comment:
It might be worth providing this as a definition, since here it's hidden away inside an `exists_unique` where it can't be used computably.

Reply:
With “this” you mean the composition of the lifts? But isn’t that just  the `is_free_group.lift` of this instance?

Reply:
Right, but annoyingly `is_free_group.lift` is itself noncomputable, for no reason really other than that we got the design wrong. Fixing that is out of scope, but might be something you're interested in trying to fix in another PR.

Reply:
What would be a better design? express `is_free_group` via an isomorphism to `free_group X`, and use the (computable) lift from there?

Reply:
The discussion is here: https://github.com/leanprover-community/mathlib/pull/7086#discussion_r608596995

Reply:
@nomeata @eric-wieser 

Do I read the above correctly that the design changes discussed here are now pegged for future work? Is this ready to be merged?

Reply:
Yes, such a refactoring is clearly out of scope for this small contribution

Reply:
Let's wait and see if Eric agrees.

Reply:
I agree

Reply:
Of course, you're more than welcome to take up that future work before this PR, but you're under no obligation to if you'd prefer this to be merged as is